### PR TITLE
Replace scanner with reader #3406

### PR DESCRIPTION
**Pull request by fritschldwg**
# Replace scanner with reader #3406

Issue: #3406

## Description

  - Fix for a bug in the piped input project checking code
  - Previously relied on a scanner to detect piped input, assuming no bytes meant no project
  - Bug occurred in interactive mode when pressing enter after the last question, causing bytes to be available
  - The scanner entered reading mode and hung indefinitely, waiting for user input
  - Resolved the issue by switching to a reader instead of a scanner
  - The reader provides a method to check whether the next read is blocking or not, preventing the hang bug
  - Also resolves point 5 of issue #3406
  
## Definition of Done

A PR is only ready for merge once all the following acceptance criteria are fulfilled:
- [x] Changes have been manually tested
- [x] All TODOs related to this PR have been closed
- [x] There are automated tests for newly written code and bug fixes
- [x] All bugs discovered while working on this PR have been submitted as issues (if not already an open issue)
- [x] Documentation (GH-pages, analysis/visualisation READMEs, parser READMEs, --help, etc.) has been updated (almost always necessary except for bug fixes)
- [x] CHANGELOG.md has been updated

### CODE REVIEW

File: analysis/model/src/main/kotlin/de/maibornwolff/codecharta/serialization/ProjectInputReader.kt [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+        val syncFlag = PipeableParserSyncFlag.SYNC_FLAG.value
+        val syncSignalBytes = syncFlag.toByteArray()
+        input.reset()
+        return isSubarray(syncSignalBytes, buffer)
+    }
+
+    private fun extractJsonObjectFromEndOfStream(streamContent: String): String {
+        var count = 0
+        val openingBracket = "{".toCharArray()[0]
```
Comment (phanlezz) [inline analysis/model/src/main/kotlin/de/maibornwolff/codecharta/serialization/ProjectInputReader.kt:71]:
```suggestion
        val openingBracket = '{'
```

**Comment by phanlezz** [comment]
Does this close #2858 ?

**Review by phanlezz — APPROVED** [review]
Closes #2858

- Void -> Unit
- Test for ProjectInputReader line 93, return streamContent if JSON malformed
- Sonar code smell interface -> fun interface
- "{" -> '{'

:rocket: lgtm!

**Comment by fritschldwg** [comment]
> Closes #2858
> 
> * Void -> Unit
> * Test for ProjectInputReader line 93, return streamContent if JSON malformed
> * Sonar code smell interface -> fun interface
> * "{" -> '{'
> 
> 🚀 lgtm!

Apparently, functional interfaces can only contain abstract functions. We could convert the PipeableParser to an abstract class, but imo this would inconsistent with the InteractiveParser. For now I am accepting the code smell.

**Pull request merged**